api/envoy/config/core/v3/socket_interface.proto

```diff
@@ -0,0 +1,25 @@
+syntax = "proto3";
+
+package envoy.config.core.v3;
+
+import "udpa/annotations/status.proto";
+import "udpa/annotations/versioning.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.config.core.v3";
+option java_outer_classname = "SocketInterfaceProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Socket Interface Configuration ]
+
+// Used to select the socket interface implementation to use
+message SocketInterfaceConfig {
+  enum SocketInterfaceType {
+    // Default, OS dependent, implementation
+    DEFAULT = 0;
```

Comment:
How do you envision this to be used? I.e. what other implementations would be defined in the public Envoy API? How is this different from using the `Envoy::Network::SocketInterfaceLoader` to override default SocketInterface?

Reply:
IMO we want this to be separate from bootstrap extensions, as there is only one of these, right? Or per @yanavlasov do we potentially need to allow multiple that are potentially looked up by the resolver?

I think my general feeling for this is that we should be specifying this by typed config, either directly via a bootstrap extension or as a discrete thing that is also specified in bootstrap?

Reply:
Given the conversation on #11618, I think that just having the ability to inject a different implementation (bootstrap or not) may be enough. The thing that we're missing is a way to signal to the custom `SocketInterface` that certain `Address::Instance`s should be treated differently. 

As for the best option to change the `SocketInterface`, I'll let you be the judge of it! :-)

My rather naive (and probably wrong) idea was to add to extensions "socket interface shims" that would link to actual implementations, have them imported in "socket_interface_factory.h" and added as options in `socket_interface.proto`.

Reply:
Should this be on the `Address` itself, similar to the resolver?

Reply:
@htuch, that's an interesting idea. My original plan was to have a way of overriding the default `SocketInterface` implementation with a startup conf* because I assumed we wanted only one such interface to be active at a time.

Are you thinking about having multiple `SocketInterfaces` in a `FactoryRegistry` and then, based on `Address` attributes, retrieve and ask the factory to generate an `IoHandle` (or a `Socket` if we later decide to merge the two)? It sounds good to me.

*side note: just did some testing and realized that bootstrap_extensions are loaded after the `ListenerManager` is initialized. So, this can't work as is.

Reply:
Yes, similar to how we deal with the resolver. One thing to be cautious of is proto bloat, since we may have a lot of addresses in an Envoy config, so having to repeat the opaque config for each extension in each address might be expensive (I think the existing resolver interface suffers from this). So, I actually think I'm more in favor of the tags that @ggreenway suggested; I see this as being somewhat similar to the transport socket matchers that are now used by clusters.

Reply:
I agree that this should be a typed config, similar to other extensions configuration. Otherwise you'd have to leak extension specific config into the core API.
I do not have a strong sense whether this should involve addresses or not. I can see someone adding iWARP support in Envoy and it might not involve addresses.

Reply:
Would then a hash_map/vector of `Address::Tag` objects (to be defined) in `Address::InstanceBase` that can be updated via `Address::Instance` apis be enough? Or do you think we need something more generic? (cc @ggrenway, @yanavlasov)

This would limit configurability but it should be pretty light weight. It does have the advantage that the resolver can easily pass opaque information to the `SocketInterface` implementation.

Reply:
@yanavlasov agreed and apologies, I somehow managed to miss your comment. 
To summarize, I guess we now have two issues that need to be solved:
- configuration of the `SocketInterface` implementation(s). Should be a typed config. I'm probably confused here, but the original goal of `socket_interface` was to use it as a typed_config under bootstrap_extensions and to keep on adding types to the enum, after `DEFAULT`, as we add more implementations to extensions. Were you thinking about something else? Even so, this doesn't work as a bootstrap extension because it's initialized after `ListenerManager`, i.e., after the addresses are bound. I guess we'll have to move it directly under boostrap. 
- passing of extra context between resolvers and the `SocketInterface` via `Address::Instance`s. My previous comment was only concerned with this.